README.md
# cyou-big-number

![Travis (.com)](https://img.shields.io/travis/com/teeeemoji/cyou-big-number)
![Codecov](https://img.shields.io/codecov/c/github/teeeemoji/cyou-big-number?token=8b1462035af64c20adae6bb75dd2cda6)
![NPM](https://img.shields.io/npm/l/cyou-big-number)
![npm](https://img.shields.io/npm/v/cyou-big-number)
![npm](https://img.shields.io/npm/dy/cyou-big-number)
![npm bundle size](https://img.shields.io/bundlephobia/min/cyou-big-number?style=plastic)
![GitHub last commit](https://img.shields.io/github/last-commit/teeeemoji/cyou-big-number)

Simple decimal arithmetic for the browser and node.js

This package is based on [decimal.js-light](decimal.js) and offer some very simple API for **big decimal**.

-   {number | string} transform to decimal
-   {decimal} to fixed format
-   decimal arithmetic add/subtract/multiply/divide

Now, see you big number.

## Installation

```console
$ npm install cyou-big-number --save
```

## Usage

### In the Browser

```html
<script src="dist/umd/index.umd.js"></script>
```

### In node

```javascript
import {numberToString, decimalToString, decimalPlus, decimalMinus, decimalTimes, decimalDiv} from 'cyou-big-number'
```

## Examples

```javascript
// Supports at least 20 integers to 10 decimals
numberToString('123456789012345678790.12345678901234567890'); // '123456789012345678790.123456789'
numberToString('9e-10'); // '0.0000000009'
numberToString(9e-10); // '0.0000000009'

// helpful format
numberToString('1.12345000000000'); // '1.12345'
numberToString('1.12345000000000', {minFixed: 10}); // '1.1234500000'
numberToString('1.12345000000000', {maxFixed: 2}); // '1.12'

// No loss of accuracy
decimalToString(decimalPlus(0.1, 0.2)); // '0.30'

decimalToString(decimalMinus(5, 1, 1, 1, 1, 1)); // '0.00'

numberToString(
    decimalTimes(Number.MAX_SAFE_INTEGER, 1000)
); // '9007199254740991000'

numberToString(
  decimalDiv(1000, 1000, 1000, 1000)
); // '0.000001'
```

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### numberToDecimalObj

[src/index.js:9-17](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L9-L17 "Source code on GitHub")

#### Parameters

-   `number`  

Returns **Decimal** 

### n2d

[src/index.js:24-24](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L24-L24 "Source code on GitHub")

Type: [numberToDecimalObj](#numbertodecimalobj)

### decimalToString

[src/index.js:38-45](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L38-L45 "Source code on GitHub")

transform String/Number/Decimal to a fixed format
 1\. minFixed &lt;= decimal part length &lt;= maxFixed
 2\. decimal part will no keep 0 at the tail, unless decimal part length is small than minFixed

#### Parameters

-   `number` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** 
-   `$1` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)**  (optional, default `{}`)
    -   `$1.minFixed`   (optional, default `2`)
    -   `$1.maxFixed`   (optional, default `10`)
-   `minFixed`  default to be 2, it means 1.0 will be transformed to 1.00, but 1.000 will not be effected
           this param will keep 0 at the tail of number
-   `maxFixed`  default to be 10, it means Math.fixed(10) and this param will no keep 0 at the tail of number,
           it means 1.11000 will transform to 1.11, bug 1.11111 will not be effected

Returns **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String))** number in format

### d2s

[src/index.js:52-52](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L52-L52 "Source code on GitHub")

This function is alias for decimalToString

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** number in string format

### numberToString

[src/index.js:59-59](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L59-L59 "Source code on GitHub")

This function is alias for decimalToString, just lik d2s

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** number in string format

### n2s

[src/index.js:66-66](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L66-L66 "Source code on GitHub")

This function is alias for numberToString

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** number in string format

### decimalPlus

[src/index.js:75-77](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L75-L77 "Source code on GitHub")

this function receive multiple addends and return the sum in a Decimal Object

#### Parameters

-   `num1` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** first addends is required
-   `nums` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** the other addends

Returns **Decimal** sum

### decimalPlusToString

[src/index.js:86-88](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L86-L88 "Source code on GitHub")

This function is similar as decimalPlus, bug return the sum in a string

#### Parameters

-   `num1` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** first addends is required
-   `nums` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** the other addends

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** sum

### decimalMinus

[src/index.js:97-99](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L97-L99 "Source code on GitHub")

this function receive one minuend and multiple subtrahend and return the subtract output in a Decimal Object

#### Parameters

-   `num1` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** first number is minuend, is required
-   `nums` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** the other subtrahend

Returns **Decimal** minuend

### decimalMinusToString

[src/index.js:108-110](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L108-L110 "Source code on GitHub")

This function is similar as decimalMinus, bug return the minuend in a string

#### Parameters

-   `num1` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** first addends is required
-   `nums` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** the other subtrahend

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** minuend

### decimalTimes

[src/index.js:119-121](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L119-L121 "Source code on GitHub")

this function receive one multiplicand and multiple multiplier, return the product in a Decimal Object

#### Parameters

-   `num1` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** first number is multiplicand, is required
-   `nums` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** the other multiplier

Returns **Decimal** the product in a Decimal Object

### decimalTimesToString

[src/index.js:130-132](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L130-L132 "Source code on GitHub")

This function is similar as decimalTimes, bug return the product in a string

#### Parameters

-   `num1` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** first multiplicand is required
-   `nums` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** the other multiplier

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** product

### decimalDiv

[src/index.js:141-143](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L141-L143 "Source code on GitHub")

this function receive one dividend and multiple divisor, return the product in a Decimal Object

#### Parameters

-   `num1` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** first number is dividend, is required
-   `nums` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** the other divisor

Returns **Decimal** the dividend in a Decimal Object

### decimalDivToString

[src/index.js:152-154](https://git@github.com/:teeeemoji/cyou-big-number/blob/05f4853e3058295c41c4b6078f8c07069413fde7/src/index.js#L152-L154 "Source code on GitHub")

This function is similar as decimalDiv, bug return the dividend in a string

#### Parameters

-   `num1` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** first dividend is required
-   `nums` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | Decimal)** the other divisor

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** dividend

## Related

-   [decimal.js][decimal.js] - Consistency Made Simple

## TODOS

-   Enables functions to support chain calls
-   add prettier budge

## License

This project is licensed under the [MIT license](LICENSE).

[decimal.js]: http://mikemcl.github.io/decimal.js/


CHANGELOG.md
<a name="1.0.6"></a>
## [1.0.6](https://github.com/teeeemoji/cyou-big-number/compare/v1.0.5...v1.0.6) (2019-11-03)


### Bug Fixes

* fix error when number is 0 ([d68ff1e](https://github.com/teeeemoji/cyou-big-number/commit/d68ff1e))



<a name="1.0.5"></a>
## [1.0.5](https://github.com/teeeemoji/cyou-big-number/compare/v1.0.4...v1.0.5) (2019-10-30)


### Bug Fixes

* remove postinstall in scripts ([c15281a](https://github.com/teeeemoji/cyou-big-number/commit/c15281a))



<a name="1.0.4"></a>
## [1.0.4](https://github.com/teeeemoji/cyou-big-number/compare/v1.0.3...v1.0.4) (2019-10-30)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/teeeemoji/cyou-big-number/compare/v1.0.2...v1.0.3) (2019-10-25)



<a name="1.0.2"></a>
## [1.0.2](https://github.com/teeeemoji/cyou-big-number/compare/v1.0.1...v1.0.2) (2019-10-23)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/teeeemoji/cyou-big-number/compare/v1.0.0...v1.0.1) (2019-10-23)



<a name="1.0.0"></a>
# [1.0.0](https://github.com/teeeemoji/cyou-big-number/compare/7235b6b...v1.0.0) (2019-10-23)


### Features

* main code ([7235b6b](https://github.com/teeeemoji/cyou-big-number/commit/7235b6b))





__test__/main.test.js
/**
 * Summary.
 *  main test file.
 *
 * Description
 *  This is the main test file
 *
 * @file   This main test file for class cyou-big-number.
 * @author teeeemoji.
 * @since 2019-10-21
 */
import {
  n2s,
  decimalPlus,
  decimalPlusToString,
  decimalMinus,
  decimalMinusToString,
  decimalTimes,
  decimalTimesToString,
  decimalDiv,
  decimalDivToString
} from '../src'

describe('Number format', function () {
  test('digitizer accuracy from 10^20 to 10^-10', function () {
    expect(
      n2s('9e-10')
    ).toBe('0.0000000009')

    expect(
      n2s(9e-10)
    ).toBe('0.0000000009')

    expect(
      n2s('123456789012345678790.12345678901234567890')
    ).toBe('123456789012345678790.123456789')

    expect(
      n2s('99999999999999999999.9999999999')
    ).toBe('99999999999999999999.9999999999')

    expect(
      n2s(0.99, {maxFixed: 0, minFixed: 0})
    ).toBe('1')

    expect(
      n2s('9999999999.999999999', {maxFixed: 2})
    ).toBe('10000000000.00')
  })
})

describe('decimalPlus', function () {
  test('1 + 2 + 3 + 4 === 10', function () {
    expect(
      n2s(
        decimalPlus(1, 2, 3, 4),
        {minFixed: 0}
      )
    ).toBe('10')
  })
  test('0.1 + 0.2 === 0.3', function () {
    expect(
      n2s(decimalPlus(0.1, 0.2),
        {minFixed: 0}
      )
    ).toBe('0.3')
  })

})
describe('decimalPlusToString', function () {
  test('1 + 2 + 3 + 4 === 10', function () {
    expect(
      decimalPlusToString(1, 2, 3, 4)
    ).toBe('10.00')
  })
  test('0.1 + 0.2 === 0.3', function () {
    expect(
      decimalPlusToString(0.1, 0.2)
    ).toBe('0.30')
  })
})

describe('decimalMinus', function () {
  test('1234567890.1234567890 - 1000000000.000000009 - 200000000.00000008 - 30000000.0000007 === 4567890.123456', function () {
    expect(
      n2s(
        decimalMinus('1234567890.1234567890', '1000000000.000000009', '200000000.00000008', '30000000.0000007'),
        {minFixed: 0}
      )
    ).toBe('4567890.123456')
  })
  test('0.4 - 0.1 === 0.3', function () {
    expect(
      n2s(
        decimalMinus(0.4, 0.1),
        {minFixed: 0}
      )
    ).toBe('0.3')
  })
})

describe('decimalMinusToString', function () {
  test('1234567890.1234567890 - 1000000000.000000009 - 200000000.00000008 - 30000000.0000007 === 4567890.123456', function () {
    expect(
      decimalMinusToString('1234567890.1234567890', '1000000000.000000009', '200000000.00000008', '30000000.0000007')
    ).toBe('4567890.123456')
  })
  test('0.4 - 0.1 === 0.3', function () {
    expect(
      decimalMinusToString(0.4, 0.1)
    ).toBe('0.30')
  })
})

describe('decimalTimes', function () {
  test('0.0000000001 * 10000000000 * 2 * 3 === 6', function () {
    expect(
      n2s(
        decimalTimes('0.0000000001', '10000000000', 2, 3),
        {minFixed: 0}
      )
    ).toBe('6')
  })
  test('Number.MAX_SAFE_INTEGER * 10 === 90071992547409910', function () {
    expect(
      n2s(
        decimalTimes(Number.MAX_SAFE_INTEGER, 10),
        {minFixed: 0}
      )
    ).toBe('90071992547409910')
  })
})


describe('decimalTimesToString', function () {
  test('0.0000000001 * 10000000000 * 2 * 3 === 6', function () {
    expect(
      decimalTimesToString('0.0000000001', '10000000000', 2, 3)
    ).toBe('6.00')
  })
  test('Number.MAX_SAFE_INTEGER * 10 === 90071992547409910', function () {
    expect(
      decimalTimesToString(Number.MAX_SAFE_INTEGER, 10)
    ).toBe('90071992547409910.00')
  })
})


describe('decimalDiv', function () {
  test('1000 / 1000 / 1000 / 1000 = 0.000001', function () {
    expect(
      n2s(
        decimalDiv(1000, 1000, 1000, 1000),
        {minFixed: 0}
      )
    ).toBe('0.000001')
  })
  test('99999999999999999999.9999999999 / 33333333333333333333.3333333333 / 2 = 1.5', function () {
    expect(
      n2s(
        decimalDiv('99999999999999999999.9999999999', '33333333333333333333.3333333333', 2),
        {minFixed: 0}
      )
    ).toBe('1.5')
  })
})

describe('decimalDivToString', function () {
  test('1000 / 1000 / 1000 / 1000 = 0.000001', function () {
    expect(
      decimalDivToString(1000, 1000, 1000, 1000)
    ).toBe('0.000001')
  })
  test('99999999999999999999.9999999999 / 33333333333333333333.3333333333 / 2 = 1.5', function () {
    expect(
      decimalDivToString('99999999999999999999.9999999999', '33333333333333333333.3333333333', 2)
    ).toBe('1.50')
  })
})





src/index.js
import Decimal from 'decimal.js-light'

/**
 * @public
 * Parse argument to Decimal instance in a safe way
 * @param number
 * @returns {Decimal}
 */
export function numberToDecimalObj(number) {
  if (number instanceof Decimal) {
    return number
  }
  if (isNaN(number)) {
    throw new Error('please input a Number, or a String of a number')
  }
  return new Decimal(number || 0)
}

/**
 * @public
 * This function is alias for numberToDecimalObj
 * @type {numberToDecimalObj}
 */
export const n2d = numberToDecimalObj

/**
 * transform String/Number/Decimal to a fixed format
 *  1. minFixed <= decimal part length <= maxFixed
 *  2. decimal part will no keep 0 at the tail, unless decimal part length is small than minFixed
 * @public
 * @param {string | number | Decimal} number
 * @param minFixed default to be 2, it means 1.0 will be transformed to 1.00, but 1.000 will not be effected
 *        this param will keep 0 at the tail of number
 * @param maxFixed default to be 10, it means Math.fixed(10) and this param will no keep 0 at the tail of number,
 *        it means 1.11000 will transform to 1.11, bug 1.11111 will not be effected
 * @returns {string | string} number in format
 */
export function decimalToString(number, {minFixed = 2, maxFixed = 10} = {}) {
  let decimalObj = n2d(number)
  let decimalFixedStr = decimalObj.toFixed(maxFixed)
  let fixed = decimalFixedStr.replace(/^\d*\.?/, '').replace(/0*$/, '').length // 这是数字真实小数位数

  fixed = Math.max(minFixed, fixed)
  return decimalObj.toFixed(fixed)
}

/**
 * This function is alias for decimalToString
 * @public
 * @return {string} number in string format
 */
export const d2s = decimalToString

/**
 * This function is alias for decimalToString, just lik d2s
 * @public
 * @return {string} number in string format
 */
export const numberToString = decimalToString

/**
 * This function is alias for numberToString
 * @public
 * @return {string} number in string format
 */
export const n2s = numberToString

/**
 * this function receive multiple addends and return the sum in a Decimal Object
 * @public
 * @param {string | number | Decimal} num1 first addends is required
 * @param {string | number | Decimal} nums the other addends
 * @returns {Decimal} sum
 */
export function decimalPlus(num1, ...nums) {
  return nums.reduce((result, num) => result.plus(n2d(num)), n2d(num1))
}

/**
 * This function is similar as decimalPlus, bug return the sum in a string
 * @public
 * @param {string | number | Decimal} num1 first addends is required
 * @param {string | number | Decimal} nums the other addends
 * @returns {string} sum
 */
export function decimalPlusToString(num1, ...nums) {
  return d2s(decimalPlus(num1, ...nums))
}

/**
 * this function receive one minuend and multiple subtrahend and return the subtract output in a Decimal Object
 * @public
 * @param {string | number | Decimal} num1 first number is minuend, is required
 * @param {string | number | Decimal} nums the other subtrahend
 * @returns {Decimal} minuend
 */
export function decimalMinus(num1, ...nums) {
  return nums.reduce((result, num) => result.minus(n2d(num)), n2d(num1))
}

/**
 * This function is similar as decimalMinus, bug return the minuend in a string
 * @public
 * @param {string | number | Decimal} num1 first addends is required
 * @param {string | number | Decimal} nums the other subtrahend
 * @returns {string} minuend
 */
export function decimalMinusToString(num1, ...nums) {
  return d2s(decimalMinus(num1, ...nums))
}

/**
 * this function receive one multiplicand and multiple multiplier, return the product in a Decimal Object
 * @public
 * @param {string | number | Decimal} num1 first number is multiplicand, is required
 * @param {string | number | Decimal} nums the other multiplier
 * @returns {Decimal} the product in a Decimal Object
 */
export function decimalTimes(num1, ...nums) {
  return nums.reduce((result, num) => result.times(n2d(num)), n2d(num1))
}

/**
 * This function is similar as decimalTimes, bug return the product in a string
 * @public
 * @param {string | number | Decimal} num1 first multiplicand is required
 * @param {string | number | Decimal} nums the other multiplier
 * @returns {string} product
 */
export function decimalTimesToString(num1, ...nums) {
  return d2s(decimalTimes(num1, ...nums))
}

/**
 * this function receive one dividend and multiple divisor, return the product in a Decimal Object
 * @public
 * @param {string | number | Decimal} num1 first number is dividend, is required
 * @param {string | number | Decimal} nums the other divisor
 * @returns {Decimal} the dividend in a Decimal Object
 */
export function decimalDiv(num1, ...nums) {
  return nums.reduce((result, num) => result.div(n2d(num)), n2d(num1))
}

/**
 * This function is similar as decimalDiv, bug return the dividend in a string
 * @public
 * @param {string | number | Decimal} num1 first dividend is required
 * @param {string | number | Decimal} nums the other divisor
 * @returns {string} dividend
 */
export function decimalDivToString(num1, ...nums) {
  return d2s(decimalDiv(num1, ...nums))
}
